File: LeetCode/Peng/86PartitionList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
//Use two head, one collect small number, one collect big number, then link them together
//Notice: the rightmost should be NULL!!!
class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        if (!head || !head->next) return head;
        ListNode* dummyLeft = new ListNode(-1);
        ListNode* dummyRight = new ListNode(-1);
        ListNode* left = dummyLeft, * right = dummyRight;
        while (head) {
            if (head->val < x) {
                left->next = head;
                left = left->next;
            } else {
                right->next = head;
                right = right->next;
            }
            head = head->next;
        }
        left->next = dummyRight->next;
        right->next = NULL;
        left = dummyLeft->next;
        delete dummyLeft;
        delete dummyRight;
        return left;
    }
};

File: LeetCode/Hao/86PartitionList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode partition(ListNode head, int x) {
        ListNode small = new ListNode(0);
        ListNode big = new ListNode(0);
        small.next = head;
        big.next = head;
        ListNode c = big;
        ListNode start = small;

        
        for (ListNode p = head; p != null; p = p.next) {
            if (p.val < x) {
                small.next = p;
                small = small.next;
            } else {
                big.next = p;
                big = big.next;
            }
        }
        big.next = null;
        small.next = c.next;
        
        return start.next;
    }
}
